### ISSUE TITLE
Strange DPO loss pattern 

### ISSUE THREAD

**Issue by puyuanOT** [huggingface/trl#800]
It looks like the log probability of both chosen& rejected trajectory is keep decreasing, which is strage.

![image](https://github.com/eric-mitchell/direct-preference-optimization/assets/121119211/a2f635a1-0874-4f90-bc19-5083a494e14b)
![image](https://github.com/huggingface/trl/assets/121119211/2d5cdf4d-d7bf-4d8f-87b9-712165fd31e8)

**Comment by i4never** [comment]
I think it's okay because the target of DPO loss is increasing the margin between chosen/rejected sample. Model can achieve this by lower rejected reward faster.

**Comment by puyuanOT** [comment]
> I think it's okay because the target of DPO loss is increasing the margin between chosen/rejected sample. Model can achieve this by lower rejected reward faster.

But it looks like the log-prob of the chosen one is also decreasing, despite the margin. Should I add an extra term to the loss function, such as NLL of the chosen response?

**Comment by i4never** [comment]
Worth trying.

**Comment by Moyhub** [comment]
> > I think it's okay because the target of DPO loss is increasing the margin between chosen/rejected sample. Model can achieve this by lower rejected reward faster.
> 
> But it looks like the log-prob of the chosen one is also decreasing, despite the margin. Should I add an extra term to the loss function, such as NLL of the chosen response?

Hi, I have a similar problem. When I was training, both chosen_reward and reject_reward decreased, but the latter decreased more, so the margin became larger. Do you have any good solutions?
![image](https://github.com/huggingface/trl/assets/44131880/acb59e20-2b0d-4c77-9831-57a04fc72644)

**Comment by gzliyu** [comment]
> > I think it's okay because the target of DPO loss is increasing the margin between chosen/rejected sample. Model can achieve this by lower rejected reward faster.
> 
> But it looks like the log-prob of the chosen one is also decreasing, despite the margin. Should I add an extra term to the loss function, such as NLL of the chosen response?

Hi! Do you have any idea on this? In my task, I tried adding a SFT-like, cross entropy loss to encourage the model to give chosen response. But the trained one turns out to be worse.

**Comment by oswen** [comment]
> > > I think it's okay because the target of DPO loss is increasing the margin between chosen/rejected sample. Model can achieve this by lower rejected reward faster.
> > 
> > 
> > But it looks like the log-prob of the chosen one is also decreasing, despite the margin. Should I add an extra term to the loss function, such as NLL of the chosen response?
> 
> Hi! Do you have any idea on this? In my task, I tried adding a SFT-like, cross entropy loss to encourage the model to give chosen response. But the trained one turns out to be worse.

if you have tryied adding sft-loss, then I think you can try DPOP（Smaug: Fixing Failure Modes of Preference Optimisation with DPO-Positive）

**Comment by 1485840691** [comment]
@qgallouedec @LeonEricsson  If OK, I would try to give PR on implementing the DPOP loss which adds an extra regularization term to the winning sample to prevent its policy log probs from deviating from reference log probs, when reference model is enabled in DPO
https://arxiv.org/pdf/2402.13228
<img width="1440" height="214" alt="Image" src="https://github.com/user-attachments/assets/996a0e36-a024-4931-9cd9-aa865573574b" /> 

The difference between the adding this term and adding extra NLL loss on chosen sample (like rDPO) is that DPOP ensures that the chosen  policy log prob is over the chosen reference log prob so the chosen pair will get reward, while NLL loss on chosen sample might not guarantee this.

Detailed code updates in DPOTrainer:
              chosen_logratios = chosen_logps.to(device) - (not self.reference_free) * ref_chosen_logps.to(device)
              **logits = logits - self.dpop_penalty_coef * torch.clamp(-chosen_logratios, min=0)**
